### ISSUE TITLE
Implement Symfony Request class in OC's request class

### ISSUE THREAD

**Issue by th3fallen** [owncloud/core#14306]
I know we literally just finished this https://github.com/owncloud/core/pull/14056

But this class is not easily (correctly)mockable without actually instantiating the class and mocking its three deps.

Its also impossible to mock its $server property since it doesn't actually exist and is fetched magically via a `__get` method which is in a lengthy switch case.

extending http://api.symfony.com/2.0/Symfony/Component/HttpFoundation/Request.html in our `\OC\AppFramework\Http\Request` would give us much more maintainablilty and reduce our maintance needs as that class would handle all the heavy lifting except our specialty cases we have to account for.

Also in the future could be used when we get around to refactoring a number of other classes.

The changes on our codes end would also be minimal from my first looks. And could be made more minimal through usages of facades.

Thoughts on this @DeepDiver1975 @karlitschek 

Please also note this class is deeply tested and used in some of php's most popular frameworks like Symfony(obviously), Laravel, Drupal(not really a framework), Silex, and a number of others I can't think of at the moment.

**Comment by DeepDiver1975** [comment]
Using symfiny's request class to implement our own IRequest interface might be an option - I honestly have to have a look at this part of symfony first.

But we require our own interface declarations in the namespace OCP.

**Comment by th3fallen** [comment]
I was suggesting to have `\OC\AppFramework\Http\Request` extend `\Symfony\Component\HttpFoundation\Request` that way our interface is still viable and we can extend it to handle our edge cases.

**Comment by LukasReschke** [comment]
Generally speaking: Good idea.

I see three options here:

1. We don't do the move.
2. We do the move post-8.1 and prepare the class already by renaming new master only functions such as `getRemoteAddr` in `getClientIp` to not have to change the public API later in a stable release.
3. We do it right now ;-)

**Comment by DeepDiver1975** [comment]
> But this class is not easily (correctly)mockable without actually instantiating the class and mocking its three deps.
> 
> Its also impossible to mock its $server property since it doesn't actually exist and is fetched magically via a __get method which is in a lengthy switch case.

I'd like to hear from @Raydiation about this - he invented that wheel :speak_no_evil: 

Let me also have a look tomorrow at the symfomy classes. THX

**Comment by th3fallen** [comment]
i like option 3 obviously :) but another option would be to do it now and do a facade to keep the old method names and add a depreciated tag to them to give third parties time to move away in the off case they are using them.

But implementing this would be a great move imho and i can only assume it would be much faster (i'll report back with profiling since im playing with it anyway) 20% rule :)

**Comment by Raydiation** [comment]
No need to do that IMHO. 

* You will break existing controllers
* You will break middleware
* It is just a simple data structure
* You dont need to mock the request, you can very simply create it by using $request = new Request(['server' => [/* your parameters */]])
* Any backwards compatible implementation will be slower
* The whole class is also covered by unit tests
* We also provide url parameters in the class, switching it out will cause some pain
* Do we need the additional the methods on it?

Most of the time you inject the values in the controller method, e.g. **public function index($param1)** which basically means you will basically never access the request object. 

TL;DR: Minimal benefits with lots of breakage and work to get it running

**Comment by th3fallen** [comment]
@Raydiation 
* i cant speak for breaking controllers but if so i would imagine it would be due to usage of magic methods to set post/get/whatever data which is a one line fix.
* middleware is fixed with a single line (i have a working rough copy atm)
* creating a new request instance forces you to mock it's three dependencies which is significantly less than ideal.
* This obviously would not be backported to older versions.
* The class would still be covered by unit tests as well as the tests provided by symfony which are significantly more exhaustive.
* this is why i said to extend symfonys Request class so we can retain our edge case methods in there such as the url parameters
* its not so much that we need the additional methods as much as this library is SIGNIFICANTLY more efficient at handling requests than the current class. It also allows much more flexibility. Most of the additional methods are ones we already have that are named differently.

I personally think in the long run this will solve many more headaches than it will cause in the short term. (has been the case in the past in my experience). But again just voicing my opinion.

**Comment by Raydiation** [comment]
Real deal breaker IMHO would be functionality.

**Comment by Raydiation** [comment]
As for middleware: I meant custom app middleware ;)

**Comment by th3fallen** [comment]
@Raydiation if it becomes a big issue that notifying them of a one line fix cannot resolve we/i/whoever could simply create a facade that could let them work as is and still use the new methods....

**Comment by th3fallen** [comment]
while noting the facade methods would be depreciated to stop future usage and encourage movement in the new direction.

**Comment by Raydiation** [comment]
Hm, it seems that there were some additions to the request class by @LukasReschke which I did not  know about yet and the code seems to be more entagled with our code base than I thought

I think changing this will actually be very hard and I doubt that we will actually gain anything from the merge. Sure the interface will be nicer and its used more but we have our own code which also has to be maintained and tested. 

Adding the symfony part will actually make things worse and it won't remove the needed dependencies if you need to mock it. Some constructors like create will even create an invalid request object which means that we'd need to use composition instead of inheritance (adapter pattern) which basically means: passing a request object into a request object and proxy the methods to the internal implementation adding yet another constructor dependency.

What we can do to make testing easier is to remove the need to pass the request and appname into the parent's controller constructor. That would further limit the need to use the request object.

What is the real benefit? Some more methods on an object that you actually don't use a lot.

PS: we need to implement yet another session interface

**Comment by LukasReschke** [comment]
> Hm, it seems that there were some additions to the request class by @LukasReschke which I did not know about yet and the code seems to be more entagled with our code base than I thought

That worse? – If you have ideas how to improve it further: Shoot it to me and I'll implement it :smiling_imp: 
(but IMHO way better than the static `OC_Request` class that we had before… at least the approach now allows us to test a huge bandwidth of things that we couldn't before)

**Comment by Raydiation** [comment]
Don't worry, I think most of the changes are actually fine and they are already published which means changing it will cause breakage 

all I've tried to say is that we need to override things with owncloud internal stuff like the webroot which would also require us to change the symfony stuff in a way that would work against the initially listed benefits.

**Comment by LukasReschke** [comment]
> Don't worry, I think most of the changes are actually fine and they are already published which means changing it will cause breakage

Nope. They are only in master. Nothing that a decent IDE couldn't fix automatically :-)

**Comment by Raydiation** [comment]
OK :)

@th3fallen can you investigate if we need to modify symfony methods based on config.php settings?

**Comment by th3fallen** [comment]
I can in the morning

**Comment by DeepDiver1975** [comment]
let's close this - we will stick with our request implementation - THX
